pkg/networkservice/common/begin/event_originator.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package begin
+
+import (
+	"context"
+
+	"github.com/edwarnicke/serialize"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+// EventOriginator provides a handle for midchain elements to call Request() or Close() against the beginning of
+// a chain
+type EventOriginator struct {
+	// retryRequest *networkservice.NetworkServiceRequest
+	executor serialize.Executor
+	req      *networkservice.NetworkServiceRequest
+	conn     *networkservice.Connection
+	opts     []grpc.CallOption
+
+	nextCtx context.Context
+
+	client networkservice.NetworkServiceClient
+	server networkservice.NetworkServiceServer
+}
+
+func newEventOriginator(nextCtx context.Context, server networkservice.NetworkServiceServer, client networkservice.NetworkServiceClient) *EventOriginator {
+	return &EventOriginator{
+		nextCtx: nextCtx,
+		client:  client,
+		server:  server,
+	}
+}
+
+/*
+Request - calls the begin.Request(ctx,request) from the beginning of the chain.
+
+Uses the networkservice.NetworkServiceRequest from the chain's last successfully completed Request()
+with networkservice.NetworkServiceRequest.Connection replaced with the Connection returned by the chain's last
+successfully completed Request() event
+
+Use:
+
+	func WithRequestMutator(parent context.Context, mutators ...RequestMutatorFunc) context.Context
+
+to mutate the default Request sent.
+
+Note: if a chain is a server chain continued by a client chain, the beginning of the chain is at the beginning of
+the server chain, even if there is a subsequent begin.NewClient() in the client chain.
+*/
+func (e *EventOriginator) Request(ctx context.Context) (conn *networkservice.Connection, err error) {
+	if e == nil {
+		return nil, errors.Errorf("EventOriginator.Request should not be calle on a nil receiver")
+	}
+	req := e.req.Clone()
```

Comment:
Looks like this would update outdated request clone if we have 2 requests processing in the same time (e.g. Refresh + update from the Client):
1. Client update comes and locks the executor.
2. Refresh happens and takes the stored Request clone.
3. Refresh starts waiting on the executor.
4. Client unlocks the executor and updates the stored Request value.
5. Refresh starts executing with the old Request clone.
6. Refresh finishes and updates the stored Request value with an old value (and so it updates it back for the whole chain on the right).

Probably it would be better to first lock the executor and only then take a Request clone.

Reply:
And so this also can lead us to the data race :)